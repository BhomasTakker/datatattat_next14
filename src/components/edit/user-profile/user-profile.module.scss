@use "../../../scss/abstracts/units/" as *;

.root {
	display: flex;
	gap: $spacingMd;

	.image {
		width: 100px;
	}
	.content {
		display: flex;
		flex-direction: column;
		gap: $spacingXs;

		.name {
		}
		.email {
		}
		.role {
			margin-top: auto;
		}
	}
}
