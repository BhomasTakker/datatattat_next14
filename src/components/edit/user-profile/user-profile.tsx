import styles from "./user-profile.module.scss";

type User = {
	name: string;
	email: string;
	image: string;
	role: string; //union of roles
};

// potentially only show detail i.e. email and role
// IF you ARE that user
export const UserProfile = ({ user }: { user: User }) => {
	const { name, email, image, role } = user;
	return (
		<article className={styles.root}>
			<img className={styles.image} src={image} alt="Me yar!" />

			<div className={styles.content}>
				<h2 className={styles.name}>{name}</h2>
				<p className={styles.email}>{email}</p>
				<p className={styles.role}>{`role:- ${role}`}</p>
			</div>
		</article>
	);
};
