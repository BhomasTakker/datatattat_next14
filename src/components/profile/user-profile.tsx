import styles from "./user-profile.module.scss";
import { IUser } from "@/types/user";

type ProfilePageProps = {
	user: Partial<IUser>;
};
// potentially only show detail i.e. email and role
// IF you ARE that user
export const UserProfile = ({ user }: ProfilePageProps) => {
	const {
		username,
		avatar,
		signin_email,
		signin_method,
		signin_name,
		role,
		_id,
	} = user;

	const restricted = (
		<div>
			<p className={styles.role}>{`user role:- ${role}`}</p>
			<p className={styles.method}>{signin_method}</p>
			<p className={styles.name}>{signin_name}</p>
			<p className={styles.email}>{signin_email}</p>
			<p className={styles.id}>{`user_id:- ${_id}`}</p>
		</div>
	);

	return (
		<article className={styles.root}>
			<img
				className={styles.image}
				src={avatar}
				alt={`${username} avatar image`}
			/>

			<div className={styles.content}>
				<h2 className={styles.name}>{username}</h2>
				{_id ? restricted : null}
			</div>
		</article>
	);
};
