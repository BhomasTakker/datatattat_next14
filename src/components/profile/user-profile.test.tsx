import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { UserProfile } from "./user-profile";

describe("UserProfile", () => {
	const mockUser = {
		username: "Jane Doe",
		email: "jane@example.com",
		avatar: "https://example.com/avatar.jpg",
	};
	const authorizedUser = {
		...mockUser,
		_id: "12345",
		role: "admin",
		signin_method: "email",
		signin_name: "Jane Doe",
		signin_email: "anEmail@emailProvider.com",
	};

	it("renders user name", () => {
		render(<UserProfile user={mockUser} />);
		expect(screen.getByText("Jane Doe")).toBeInTheDocument();
	});

	it("renders user avatar", () => {
		render(<UserProfile user={mockUser} />);
		const avatar = screen.getByAltText(/Jane Doe/);
		expect(avatar).toBeInTheDocument();
		expect(avatar).toHaveAttribute("src", mockUser.avatar);
	});

	it("does not render restricted info without authorization", () => {
		const unauthorizedUser = {
			...authorizedUser,
			_id: undefined,
		};
		render(<UserProfile user={unauthorizedUser} />);
		expect(screen.queryByText(unauthorizedUser.role)).not.toBeInTheDocument();
	});

	it("renders restricted info when user is authorized", () => {
		// @ts-expect-error _id is not a string
		render(<UserProfile user={authorizedUser} />);

		// second render
		expect(screen.queryByText("user role:- admin")).toBeInTheDocument();
		expect(screen.queryByText("email")).toBeInTheDocument();
		expect(screen.queryAllByText("Jane Doe")).toHaveLength(2);
		expect(screen.queryByText("user_id:- 12345")).toBeInTheDocument();
	});

	describe("Snaps", () => {
		it("unauthorized user matches snapshot", () => {
			const { asFragment } = render(<UserProfile user={mockUser} />);
			expect(asFragment()).toMatchSnapshot();
		});
		it("authorized user matches snapshot", () => {
			// @ts-expect-error _id is not a string
			const { asFragment } = render(<UserProfile user={authorizedUser} />);
			expect(asFragment()).toMatchSnapshot();
		});
	});
});
