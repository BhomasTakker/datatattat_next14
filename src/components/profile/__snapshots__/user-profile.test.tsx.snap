// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`UserProfile Snaps authorized user matches snapshot 1`] = `
<DocumentFragment>
  <article
    class="root"
  >
    <img
      alt="Jane Doe avatar image"
      class="image"
      src="https://example.com/avatar.jpg"
    />
    <div
      class="content"
    >
      <h2
        class="name"
      >
        Jane Doe
      </h2>
      <div>
        <p
          class="role"
        >
          user role:- admin
        </p>
        <p
          class="method"
        >
          email
        </p>
        <p
          class="name"
        >
          Jane Doe
        </p>
        <p
          class="email"
        >
          anEmail@emailProvider.com
        </p>
        <p
          class="id"
        >
          user_id:- 12345
        </p>
      </div>
    </div>
  </article>
</DocumentFragment>
`;

exports[`UserProfile Snaps unauthorized user matches snapshot 1`] = `
<DocumentFragment>
  <article
    class="root"
  >
    <img
      alt="Jane Doe avatar image"
      class="image"
      src="https://example.com/avatar.jpg"
    />
    <div
      class="content"
    >
      <h2
        class="name"
      >
        Jane Doe
      </h2>
    </div>
  </article>
</DocumentFragment>
`;
